### ISSUE TITLE
Does blocking metadata server ip cause istio-proxy sidecar container to fail

### ISSUE THREAD

**Issue by seastranger** [istio/istio#40010]
Hello,
we are usinng istio 1.13.5 in our clusters.
in one of the clusters in GCP we see behavior, where pods with configured istio-proxy sidecars (for mtls) fail to start with readiness check failing:
```
Warning FailedPostStartHook pod/workspaces-ws-7q8dm-deployment-659c5b85cf-6kjrl Exec lifecycle hook ([pilot-agent wait]) 
for Container "istio-proxy" in Pod "workspaces-ws-7q8dm-deployment-659c5b85cf-6kjrl_ws-ns-devoted-wildcat(c9614831-4e43-4ff0-b7be-a5264d97a5f4)" failed - 
error: command 'pilot-agent wait' exited with 255: Error: timeout waiting for Envoy proxy to become ready. 
Last error: Get "[http://localhost:15021/healthz/ready":](http://localhost:15021/healthz/ready%22:) dial tcp [::1]:15021: connect: connection refused...
```
when checking istio-proxy container startup logs we see the following:
```
Jul 17, 2022 @ 19:42:44.610	2022-07-17T16:42:44.609890Z	error	accept tcp [::]:15020: use of closed network connection
Jul 17, 2022 @ 19:41:44.515	2022-07-17T16:41:44.514245Z	info	cache	Root cert has changed, start rotating root cert
Jul 17, 2022 @ 19:41:44.515	2022-07-17T16:41:44.514267Z	info	ads	XDS: Incremental Pushing:0 ConnectedEndpoints:0 Version:
Jul 17, 2022 @ 19:41:44.515	2022-07-17T16:41:44.514325Z	info	cache	returned workload trust anchor from cache	ttl=23h59m59.485679302s
Jul 17, 2022 @ 19:41:44.514	2022-07-17T16:41:44.514195Z	info	cache	generated new workload certificate	latency=175.923907ms ttl=23h59m59.485822392s
Jul 17, 2022 @ 19:41:44.341	2022-07-17T16:41:44.340962Z	info	sds	Starting SDS grpc server
Jul 17, 2022 @ 19:41:44.337	2022-07-17T16:41:44.337216Z	info	ads	All caches have been synced up in 45.497274ms, marking server ready
Jul 17, 2022 @ 19:41:44.337	2022-07-17T16:41:44.337475Z	info	sds	SDS server for workload certificates started, listening on "etc/istio/proxy/SDS"
Jul 17, 2022 @ 19:41:44.337	2022-07-17T16:41:44.337507Z	info	xdsproxy	Initializing with upstream address "istiod.istio-system.svc:15012" and cluster "Kubernetes"
Jul 17, 2022 @ 19:41:44.310	2022-07-17T16:41:44.310119Z	info	citadelclient	Citadel client using custom root cert: var/run/secrets/istio/root-cert.pem
Jul 17, 2022 @ 19:41:44.302	2022-07-17T16:41:44.302316Z	error	error in getting aws info for iam/info : Get "http://[fd00:ec2::254]/latest/meta-data/iam/info": dial tcp [fd00:ec2::254]:80: connect: network is unreachable
```
now in the pod's namespace we have a networkpolicy that blocks access to metadata server ip 169.254.169.254 (for security reasons), which aparently causes the network unreachable error above, when we remove this rule container starts successfully.

Is this the expected behavior ? does istio-proxy container requires access to metadata server, and fails to start without it ?
Thanks in advance for your help.

Sergey

**Comment by howardjohn** [comment]
cc @zirain

**Comment by zirain** [comment]
@seastranger can you share how you install istio? the iop file or values.yaml for helm chart.

**Comment by seastranger** [comment]
here you go:
```
apiVersion: v1
items:
- apiVersion: install.istio.io/v1alpha1
  kind: IstioOperator
  metadata:
    annotations:
      helm.sh/hook-weight: "1"
    finalizers:
    - istio-finalizer.install.istio.io
    generation: 3
    labels:
      app.kubernetes.io/instance: istio
    name: appstudio
    namespace: istio-system
  spec:
    components:
      ingressGateways:
      - k8s:
          affinity:
            podAntiAffinity:
              preferredDuringSchedulingIgnoredDuringExecution:
              - podAffinityTerm:
                  labelSelector:
                    matchLabels:
                      app: istio-ingressgateway
                  topologyKey: topology.kubernetes.io/zone
                weight: 100
              requiredDuringSchedulingIgnoredDuringExecution:
              - labelSelector:
                  matchLabels:
                    app: istio-ingressgateway
                topologyKey: kubernetes.io/hostname
          hpaSpec:
            maxReplicas: 3
            minReplicas: 3
          resources:
            limits:
              cpu: "3"
              memory: 20Gi
            requests:
              cpu: 50m
              memory: 384Mi
        name: istio-ingressgateway
      pilot:
        k8s:
          affinity:
            podAntiAffinity:
              preferredDuringSchedulingIgnoredDuringExecution:
              - podAffinityTerm:
                  labelSelector:
                    matchLabels:
                      app: istiod
                  topologyKey: topology.kubernetes.io/zone
                weight: 100
              requiredDuringSchedulingIgnoredDuringExecution:
              - labelSelector:
                  matchLabels:
                    app: istiod
                topologyKey: kubernetes.io/hostname
          env:
          - name: PILOT_PUSH_THROTTLE
            value: "1"
          - name: PILOT_DEBOUNCE_AFTER
            value: 10s
          hpaSpec:
            maxReplicas: 3
            minReplicas: 2
          resources:
            limits:
              cpu: "3"
              memory: 15Gi
            requests:
              cpu: 10m
              memory: 128Mi
    meshConfig:
      accessLogFile: /dev/stdout
      accessLogFormat: |
        [%START_TIME%] "%REQ(:METHOD)% %REQ(X-ENVOY-ORIGINAL-PATH?:PATH)% %PROTOCOL% %DOWNSTREAM_TLS_VERSION%" %RESPONSE_CODE% %RESPONSE_FLAGS% %RESPONSE_CODE_DETAILS% %CONNECTION_TERMINATION_DETAILS%"%UPSTREAM_TRANSPORT_FAILURE_REASON%" %BYTES_RECEIVED% %BYTES_SENT% %DURATION% %RESP(X-ENVOY-UPSTREAM-SERVICE-TIME)% "%REQ(X-FORWARDED-FOR)%" "%REQ(USER-AGENT)%" "%REQ(X-REQUEST-ID)%""%REQ(:AUTHORITY)%" "%UPSTREAM_HOST%" %UPSTREAM_CLUSTER% %UPSTREAM_LOCAL_ADDRESS% %DOWNSTREAM_LOCAL_ADDRESS% %DOWNSTREAM_REMOTE_ADDRESS% %REQUESTED_SERVER_NAME% %ROUTE_NAME%
      rootNamespace: istio-root
    profile: default
    values:
      global:
        hub: build-releases-external.common.cdn.repositories.cloud.sap/istio
        imagePullSecrets:
        - istio-pull-secret
        logging:
          level: default:error
        proxy:
          holdApplicationUntilProxyStarts: true
          logLevel: error
      sidecarInjectorWebhook:
        enableNamespacesByDefault: false
        objectSelector:
          autoInject: false
```

**Comment by zirain** [comment]
@seastranger I can not reproduce this with following config:
```yaml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: test-network-policy
spec:
  podSelector: {}
  policyTypes:
  - Egress
  egress:
  - to:
    - ipBlock:
        cidr: 0.0.0.0/0
        except:
          - 169.254.169.254/32
```

there's something odd in your logs:
> Jul 17, 2022 @ 19:42:44.610	2022-07-17T16:42:44.609890Z	error	accept tcp [::]:15020: use of closed network connection

do you block port 15020?

**Comment by seastranger** [comment]
Not as far as i know
here is excerpt from our networkpolicy:
```
  spec:
    egress:
    - to:
      - ipBlock:
          cidr: 0.0.0.0/0
          except:
          - 100.96.0.0/11
          - 169.254.0.0/16
          - 100.64.0.0/13
          - 10.250.0.0/16
```
if we remove the  - 169.254.0.0/16 range, container successfully starts

**Comment by seastranger** [comment]
specifically events we see in the describe of failing pod:
```
 Warning  FailedPostStartHook  16s   kubelet             Exec lifecycle hook ([pilot-agent wait]) for Container "istio-proxy" in Pod "workspaces-ws-fpfmb-deployment-bcbd94dd6-7ppzl_ws-ns-rare-wallaby(284a379b-aeef-465d-8a02-ee3571d4f570)" failed - error: command 'pilot-agent wait' exited with 255: Error: timeout waiting for Envoy proxy to become ready. Last error: Get "http://localhost:15021/healthz/ready": dial tcp [::1]:15021: connect: connection refused
, message: "2022-07-24T10:34:07.814168Z\tinfo\tWaiting for Envoy proxy to be ready (timeout: 60 seconds)...\n2022-07-24T10:35:07.951570Z\terror\ttimeout waiting for Envoy proxy to become ready. Last error: Get \"http://localhost:15021/healthz/ready\": dial tcp [::1]:15021: connect: connection refused\nError: timeout waiting for Envoy proxy to become ready. Last error: Get \"http://localhost:15021/healthz/ready\": dial tcp [::1]:15021: connect: connection refused\n"
  Normal  Killing  16s  kubelet  FailedPostStartHook
```

**Comment by zirain** [comment]
> specifically events we see in the describe of failing pod:
> 
> ```
>  Warning  FailedPostStartHook  16s   kubelet             Exec lifecycle hook ([pilot-agent wait]) for Container "istio-proxy" in Pod "workspaces-ws-fpfmb-deployment-bcbd94dd6-7ppzl_ws-ns-rare-wallaby(284a379b-aeef-465d-8a02-ee3571d4f570)" failed - error: command 'pilot-agent wait' exited with 255: Error: timeout waiting for Envoy proxy to become ready. Last error: Get "http://localhost:15021/healthz/ready": dial tcp [::1]:15021: connect: connection refused
> , message: "2022-07-24T10:34:07.814168Z\tinfo\tWaiting for Envoy proxy to be ready (timeout: 60 seconds)...\n2022-07-24T10:35:07.951570Z\terror\ttimeout waiting for Envoy proxy to become ready. Last error: Get \"http://localhost:15021/healthz/ready\": dial tcp [::1]:15021: connect: connection refused\nError: timeout waiting for Envoy proxy to become ready. Last error: Get \"http://localhost:15021/healthz/ready\": dial tcp [::1]:15021: connect: connection refused\n"
>   Normal  Killing  16s  kubelet  FailedPostStartHook
> ```
> 
> which basically means that istio-proxy is not listening to health api

can you share the pod yaml and logs of istio-proxy container?

**Comment by zirain** [comment]
maybe same as https://github.com/istio/istio/issues/39950, hope can be fixed by https://github.com/istio/istio/pull/40105

**Comment by rinormaloku** [comment]
@zirain the PR #40105 is merged that logs an error when the metadata endpoint is not accessible and executes the requests concurrently. This fixes this issue which is the same as #39950

**Comment by zirain** [comment]
fixed by https://github.com/istio/istio/pull/40105
